test/integration/upstreams/per_host_upstream_request.cc

```diff
@@ -0,0 +1,50 @@
+#include "test/integration/upstreams/per_host_upstream_request.h"
+
+#include "common/router/router.h"
+
+using Envoy::Router::GenericConnectionPoolCallbacks;
+
+namespace Envoy {
+
+void PerHostHttpConnPool::newStream(GenericConnectionPoolCallbacks* callbacks) {
+  callbacks_ = callbacks;
+  // It's possible for a reset to happen inline within the newStream() call. In this case, we
+  // might get deleted inline as well. Only write the returned handle out if it is not nullptr to
+  // deal with this case.
+  Envoy::Http::ConnectionPool::Cancellable* handle =
+      conn_pool_->newStream(callbacks->upstreamToDownstream(), *this);
+  if (handle) {
+    conn_pool_stream_handle_ = handle;
+  }
+}
+
+bool PerHostHttpConnPool::cancelAnyPendingStream() {
+  if (conn_pool_stream_handle_) {
+    conn_pool_stream_handle_->cancel(ConnectionPool::CancelPolicy::Default);
+    conn_pool_stream_handle_ = nullptr;
+    return true;
+  }
+  return false;
+}
+
+absl::optional<Envoy::Http::Protocol> PerHostHttpConnPool::protocol() const {
+  return conn_pool_->protocol();
+}
+
+void PerHostHttpConnPool::onPoolFailure(ConnectionPool::PoolFailureReason reason,
+                                        absl::string_view transport_failure_reason,
+                                        Upstream::HostDescriptionConstSharedPtr host) {
+  callbacks_->onPoolFailure(reason, transport_failure_reason, host);
```

Comment:
presumably we should reset the handle ptr here as well?

Reply:
set   conn_pool_stream_handle_  to nullptr here and add comment to conn_pool_stream_handle_ declaration